File: src/main/golang/leetcode/editor/cn/[18]4Sum.go
package lc

import "sort"

//leetcode submit region begin(Prohibit modification and deletion)
func fourSum(nums []int, target int) [][]int {
	sort.Ints(nums)
	n := len(nums)
	findTarget := func(start, target int) (ans [][]int) {
		l, r := start, n-1
		for l < r {
			left, right := nums[l], nums[r]
			sum := left + right
			if sum < target {
				for l < r && nums[l] == left {
					l++
				}
			} else if sum > target {
				for l < r && nums[r] == right {
					r--
				}
			} else {
				ans = append(ans, []int{left, right})
				for l < r && nums[l] == left {
					l++
				}
				for l < r && nums[r] == right {
					r--
				}
			}
		}
		return
	}

	var ans [][]int
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			res := findTarget(j+1, target-nums[i]-nums[j])
			for _, item := range res {
				ans = append(ans, []int{nums[i], nums[j], item[0], item[1]})
			}
			for j < n-2 && nums[j+1] == nums[j] {
				j++
			}
		}
		for i < n-3 && nums[i+1] == nums[i] {
			i++
		}
	}
	return ans
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[18]4Sum.py
from typing import List


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        ans = []
        for i in range(n - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            for j in range(i + 1, n - 2):
                if j != i + 1 and nums[j] == nums[j - 1]:
                    continue
                l, r = j + 1, n - 1
                while l < r:
                    left, right = nums[l], nums[r]
                    sum = nums[i] + nums[j] + left + right
                    if sum == target:
                        ans.append([nums[i], nums[j], nums[l], nums[r]])
                        while l < r and nums[l] == left:
                            l += 1
                        while l < r and nums[r] == right:
                            r -= 1
                    elif sum < target:
                        while l < r and nums[l] == left:
                            l += 1
                    else:
                        while l < r and nums[r] == right:
                            r -= 1
        return ans
# leetcode submit region end(Prohibit modification and deletion)
